### ISSUE TITLE
process_info(Pid, memory) scales poorly with message queue length

### ISSUE THREAD

**Issue by dszoboszlay** [erlang/otp#6494]
**Describe the bug**
We have experienced serious performance degradation caused by a monitoring tool on a busy `gen_statem` process. The monitoring tool periodically queries some `process_info` items, but it has trouble when the `gen_statem` builds up a long message queue. A sample call (made manually, the monitoring tool queries all the items in a single call):
```erlang
1> [timer:tc(erlang, process_info, [P, [T]]) || T <- [reductions, memory, message_queue_len, current_function]]. 
[{13,[{reductions,1513399123}]},
 {1351939,[{memory,2113128516}]},
 {167,[{message_queue_len,10270709}]},
 {5,[{current_function,{gen,do_call,4}}]}]
```
At this point the process had 10M messages in its mailbox, and used ~2GB memory (which is essentially the mailbox size, the heap & stack were small). The problem is that querying the `memory` takes 1300 ms, it really shouldn't be that long.

**To Reproduce**
I couldn't reproduce the 1000+ ms times I've seen on the production machine, but it's quite easy to confirm that `process_info(Pid, memory)` scales linearly with the number of messages in the queue, even for an idle process. Interestingly, I was expecting worse performance when using `process_flag(message_queue_data, off_heap)`, but in fact it only has minor impact:

| message_queue_len | off_heap | on_heap |
|---|---|---|
| 1 | 6 us | 6 us |
| 10 | 7 us | 3 us |
| 100 | 5 us | 16 us |
| 1k | 26 us | 23 us |
| 10k | 259 us | 87 us |
| 100k | 847 us | 770 us |
| 1M | 8021 us | 8120 us |
| 10M | 90538 us | 89051 us |

For what it worths, the production machine/process differs in the following ways from my test environment:
* It has more cores: `smp:96:48` (and a lot more memory too)
* It stores messages `off_heap`
* It has mostly `{epgsql, pid(), {x_log_data, integer(), integer(), binary()}}` messages in its mailbox (sent from a single process)
* It's typically scheduled out while doing a `gen_server:call`

I don't know which one of these may be relevant.

**Expected behavior**
I would expect `process_info(Pid, memory)` to run in constant time. Wouldn't it be possible to cache somewhere in the process data structure the total memory used by the mailbox? This data has to be available at the time when the message is sent.

**Affected versions**
23.4.18 and 24.3.4.6

**Additional context**
This is the module I used for getting the numbers in the above table:
```erlang
-module(msgq).

-export([test/0, test/2, worker/1]).

-define(NS, [1, 10, 100, 1000, 10_000, 100_000, 1_000_000, 10_000_000]).
-define(ITEMS, [message_queue_len, heap_size, stack_size, total_heap_size, memory, current_function]).
-define(SAMPLES, 5).

test() ->
  [test(N, MsgQueue) || N <- ?NS, MsgQueue <- [off_heap, on_heap]].

test(N, MsgQueue) ->
  Pid = spawn(?MODULE, worker, [MsgQueue]),
  build_msg_queue(Pid, N),
  Res = [{Item, measure(Pid, Item)} || Item <- ?ITEMS],
  Pid ! quit,
  { N
  , MsgQueue
  , maps:from_list(Res)
  }.

worker(MsgQueue) ->
  process_flag(message_queue_data, MsgQueue),
  receive quit -> ok end.

build_msg_queue(_Pid, 0) -> ok;
build_msg_queue(Pid, N) ->
  Pid ! {req, erlang:unique_integer()},
  build_msg_queue(Pid, N - 1).

measure(Pid, Item) ->
  Samples = measure(Pid, Item, ?SAMPLES),
  lists:nth(?SAMPLES div 2, lists:sort(Samples)).

measure(_Pid, _Item, 0) -> [];
measure(Pid, Item, N) ->
  {Time, _} = timer:tc(erlang, process_info, [Pid, Item]),
  [Time | measure(Pid, Item, N - 1)].
```

**Comment by dszoboszlay** [comment]
> This operation is currently not yielding, so the scheduler executing this will be hogged by the `memory` operation for a very long time.

That's indeed a problem, which I haven't even considered. This time we were lucky with the scheduler lock-ups, but it should be fixed, because they are very likely to cause all kinds of trouble in a system.

> I'm quite sure that we haven't given any indications that this operation is fast.
>
> Some operations, like this one, are expensive which I don't see as a problem as such.

You haven't given any indications that this operation can be slow either, and the docs tend to describe potential performance issues (for example `erlang:statistics/1` does go into quite some detail about how similar statistics differ from each other in performance). Furthermore, I do see a problem with these operations being slow: `memory` is a very basic, but quite useful process metric that every sane monitoring tool would collect. And there isn't a fast alternative either:

* `stack_size` and `heap_size` would give a partial, potentially misleading picture only.
* `total_heap_size` doesn't include messages for `off_heap` processes, and for `on_heap` processes it's just as slow as `memory`.
* `message_queue_len` is valuable information, but doesn't reflect memory use.

> It would be relatively easy to make the memory operation O(1) by keeping track of the currently allocated amount of memory by updating a counter at every change. This would however impact performance for all processes and increase the size of the process structure of all processes regardless of whether you are interested in this information or not. [...] Even though the cost performance and memory wise would not be large, I'm not fond of giving this penalty to every process in the system.

I understand this argument. On a memory-constrained system the current `memory` performance is probably less of an issue (you can't really have 10 M messages in a mailbox when you have 32 MB RAM in total), but for a server 3 words per process won't be an issue: even with 262144 processed (the default `process_limit`) that would be a total of 6 MB.

So couldn't this become a configuration option? So everybody could decide for themselves whether they need smaller process data structures or constant time monitoring operations.

> The most of the performance impact would hit signaling and would especially hit the case where you've acquired the message queue lock.

Maybe this could be mitigated by using an atomic counter for the total value? It would be OK to update the total without holding a lock, for monitoring purposes it doesn't matter to update the total and the actual process structures atomically.

**Comment by mikpe** [comment]
I agree with @rickard-green that the VM shouldn't punish each and every process and message send just for this case. The monitoring tool should avoid this query when it might be problematic. I'll take a look doing so in `system_monitor` next week @dszoboszlay .

**Comment by rickard-green** [comment]
We will fix the yielding issue, but not implement an O(1) version of `process_info(_, memory)`. We, however, currently cannot say when the yielding issue will be fixed. We'll keep this issue open until it has been fixed.
